consensus/pacemaker/module.go

```diff
@@ -243,6 +265,82 @@ func (m *pacemaker) NewHeight() {
 		)
 }
 
+// This is called each time the system wants a delayed signal to build a block
+func (m *pacemaker) RegisterMinBlockTimeDelay() {
+	// Discard any previous timer
+	// DISCUSS: This should not happen, Identify cases where an active timer has to be discarded, remove cancellation logic if none
+	if m.prepareStepDelayer.cancelFunc != nil {
+		m.prepareStepDelayer.cancelFunc()
+	}
+
+	m.prepareStepDelayer.blockProposed = false
+	m.prepareStepDelayer.deadlinePassed = false
+
+	minBlockTime := time.Duration(m.pacemakerCfg.MinBlockTimeMsec * uint64(time.Millisecond))
+	ctx, cancel := context.WithCancel(context.TODO())
+	m.prepareStepDelayer.cancelFunc = cancel
+
+	go func() {
+		select {
+		case <-ctx.Done():
+			return
+		case <-m.GetBus().GetRuntimeMgr().GetClock().After(minBlockTime):
+			m.prepareStepDelayer.m.Lock()
+			defer m.prepareStepDelayer.m.Unlock()
+
+			// After the timeout, if there was any candidate request waiting for a signal, tell it to build the block
+			if m.prepareStepDelayer.ch != nil {
+				m.prepareStepDelayer.ch <- true
+				close(m.prepareStepDelayer.ch)
+				m.prepareStepDelayer.blockProposed = true
+			}
+
+			// From now on, build the block ASAP
+			m.prepareStepDelayer.deadlinePassed = true
+
+			// No need to cancel the context anymore
+			m.prepareStepDelayer.cancelFunc = nil
+		}
+	}()
+}
+
+// This is called when conditions are met by the leader to build a block but still needs to wait for the MinBlockTimeMsec delay before reaping the mempool
+// With MinBlockTimeMsec delay, multiple concurrent calls may happen
+// It makes sure that:
+//   - Block proposal is made by only one of the possible `HotstuffLeaderMessageHandler.HandleNewRoundMessage()` concurrent (because delayed) calls
+//   - If the timer expires, the first call to this method will trigger the block proposal
+//   - If a late message is received AFTER the a block is marked as proposed by another call, the late message is discarded
+//   - Reads and affectations to pacemaker.DelayBlockPreparation state are protected by a mutex
+func (m *pacemaker) DelayBlockPreparation() bool {
+	m.prepareStepDelayer.m.Lock()
+
```

Comment:
Why not just execute `defer m.prepareStepDelayer.m.Unlock()` here?

Reply:
At the end of this function
```go
func (m *pacemaker) DelayBlockPreparation() bool {
  ...
  m.prepareStepDelayer.ch = ch
  // We cannot defer the unlock here because the channel read is blocking
  m.prepareStepDelayer.m.Unlock()

  return <-ch
}
```
 we have `return <-ch` which will be blocking the call and preventing the `defer ...m.Unlock()` from being called  IIUC. The desired behavior is to `Unlock` right after the last assignment made to `m.prepeareStepDelayer`

Reply:
> we have return <-ch which will be blocking the call and preventing the defer ...m.Unlock() from being called IIUC.
> The desired behavior is to Unlock right after the last assignment made to m.prepeareStepDelayer

These are super valuable comments. Can you add them to the code?